ambry-server/src/integration-test/java/com/github/ambry/server/VcrAutomationTest.java

```diff
@@ -0,0 +1,293 @@
+
+    StrictMatchExternalViewVerifier helixBalanceVerifier =
+        new StrictMatchExternalViewVerifier(zkConnectString, vcrClusterName,
+            Collections.singleton(VcrTestUtil.helixResource), null);
+
+    // VCR cluster helix setup done.
+
+    VerifiableProperties verifiableProperties = new VerifiableProperties(props);
+    ClusterMapConfig clusterMapConfig = new ClusterMapConfig(verifiableProperties);
+    HelixClusterAgentsFactory helixClusterAgentsFactory = new HelixClusterAgentsFactory(clusterMapConfig, null, null);
+    VcrServer vcrServer =
+        new VcrServer(verifiableProperties, helixClusterAgentsFactory, null, new NoOpCloudFactory(), null);
+    vcrServer.startup();
+
+    makeSureHelixBalance(vcrServer, helixBalanceVerifier);
+    Assert.assertTrue("Partition assignment is not correct.", TestUtils.checkAndSleep(partitionCount,
+        () -> vcrServer.getVcrClusterParticipant().getAssignedPartitionIds().size(), 5000));
+    // vcr server start up done.
+
+    // Partition add case:
+    testPartitionLayout = constructInitialPartitionLayoutJSON(testHardwareLayout, partitionCount + 1, null);
+    Utils.writeJsonObjectToFile(testPartitionLayout.getPartitionLayout().toJSONObject(), partitionLayoutPath);
+
+    HelixBootstrapUpgradeUtil.bootstrapOrUpgrade(hardwareLayoutPath, partitionLayoutPath, zkLayoutPath, clusterPrefix,
+        dcName, 10, false, false, new HelixAdminFactory(), false, mainClusterStateModelDef,
+        HelixBootstrapUpgradeUtil.HelixAdminOperation.BootstrapCluster, dataNodeConfigSourceType, false);
+
+    makeSureHelixBalance(vcrServer, helixBalanceVerifier);
+    Assert.assertTrue("Partition assignment is not correct.", TestUtils.checkAndSleep(partitionCount + 1,
+        () -> vcrServer.getVcrClusterParticipant().getAssignedPartitionIds().size(), 5000));
+
+    // Partition remove case:
+
+    testPartitionLayout = constructInitialPartitionLayoutJSON(testHardwareLayout, partitionCount, null);
+    Utils.writeJsonObjectToFile(testPartitionLayout.getPartitionLayout().toJSONObject(), partitionLayoutPath);
+    HelixBootstrapUpgradeUtil.bootstrapOrUpgrade(hardwareLayoutPath, partitionLayoutPath, zkLayoutPath, clusterPrefix,
+        dcName, 10, false, true, new HelixAdminFactory(), false, mainClusterStateModelDef,
+        HelixBootstrapUpgradeUtil.HelixAdminOperation.BootstrapCluster, dataNodeConfigSourceType, false);
+
+    makeSureHelixBalance(vcrServer, helixBalanceVerifier);
+    Assert.assertTrue("Partition assignment is not correct.", TestUtils.checkAndSleep(partitionCount,
+        () -> vcrServer.getVcrClusterParticipant().getAssignedPartitionIds().size(), 5000));
+
+    helixControllerManager.syncStop();
+    vcrHelixControllerManager.syncStop();
+    zkInfoList.get(0).shutdown();
+  }
+
+  class NoOpCloudFactory implements CloudDestinationFactory {
+
+    @Override
+    public CloudDestination getCloudDestination() throws IllegalStateException {
+      return new CloudDestination() {
```

Comment:
You can likely use mockito here since it doesn't seem to be a complicated mock.

Reply:
fixed.